lib/src/modules/task.cc

```diff
@@ -0,0 +1,320 @@
+          },
+          "uri": {
+            "type": "object",
+            "properties": {
+              "path": {
+                "type": "string"
+              },
+              "params": {
+                 "type": "object"
+              }
+            },
+            "required": ["path", "params"]
+          },
+          "sha256": {
+            "type": "string"
+          }
+        },
+        "required": ["filename", "uri", "sha256"]
+      },
+      "minItems": 1
+    },
+    "input": {
+      "type": "object"
+    },
+    "inputFormat": {
+      "type": "string"
+    }
+  },
+  "required": ["task", "files", "input"]
+}
+)" };
+
+#ifdef _WIN32
+// The extension is required with lth_exec::execute when using a full path.
+static const std::string TASK_WRAPPER_EXECUTABLE { "task_wrapper.exe" };
+#else
+static const std::string TASK_WRAPPER_EXECUTABLE { "task_wrapper" };
+#endif
+static const int TASK_WRAPPER_FILE_ERROR_EC { 5 };
+
+// Hard-code interpreters on Windows. On non-Windows, we still rely on permissions and #!
+static const std::map<std::string, std::function<TaskCommand(std::string)>> BUILTIN_TASK_INTERPRETERS {
+#ifdef _WIN32
+    {".rb",  [](std::string filename) { return TaskCommand {
+        "ruby", { filename }
+    }; }},
+    {".pp",  [](std::string filename) { return TaskCommand {
+        "puppet", { "apply", filename }
+    }; }},
+    {".ps1", [](std::string filename) { return TaskCommand {
+        "powershell", { "-NoProfile", "-NonInteractive", "-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filename }
+    }; }}
+#endif
+};
+
+Task::Task(const fs::path& exec_prefix,
+           const std::string& task_cache_dir,
+           const std::string& spool_dir) :
+    storage_ { spool_dir },
+    task_cache_dir_ { task_cache_dir },
+    wrapper_executable_ { (exec_prefix / TASK_WRAPPER_EXECUTABLE).string() }
+{
+    module_name = "task";
+    actions.push_back(TASK_RUN_ACTION);
+
+    PCPClient::Schema input_schema { TASK_RUN_ACTION, TASK_RUN_ACTION_INPUT_SCHEMA };
+    PCPClient::Schema output_schema { TASK_RUN_ACTION };
+
+    input_validator_.registerSchema(input_schema);
+    results_validator_.registerSchema(output_schema);
+}
+
+static void addParametersToEnvironment(const lth_jc::JsonContainer &input, std::map<std::string, std::string> &environment)
+{
+    for (auto& k : input.keys()) {
+        auto val = input.type(k) == lth_jc::String ? input.get<std::string>(k) : input.toString(k);
+        environment.emplace("PT_"+k, move(val));
+    }
+}
+
+static TaskCommand getTaskCommand(const fs::path &task_executable)
+{
+    auto builtin = BUILTIN_TASK_INTERPRETERS.find(task_executable.extension().string());
+    if (builtin != BUILTIN_TASK_INTERPRETERS.end()) {
+        return builtin->second(task_executable.string());
+    }
+
+    return TaskCommand { task_executable.string(), { } };
+}
+
+// Verify (this includes checking the SHA256 checksums) that *all* task files are present
```

Comment:
Since Orchestrator is initially expected to send only a single item in `files`, this is mostly unused functionality. But it's really easy to do a loop to ensure all files are downloaded. Not sure I see a backdoor being added. We can discuss this more as part of PCP-769.